Sylvia whittle/145 dev dev remove gwyddion parity

SylviaWhittle opened the pull request:
Resolves #145 
Resolves #143 
Resolves #142 

This makes the necessary changes to achieve parity between `dev` and `dev-remove-gwyddion`, and thus by extension, `master`. I believe this change means that the output statistics of `grainstats.py` are accurate and can be considered usable. 

Here is evidence of the parity:

Before the changes, as shown in #145:
![image](https://user-images.githubusercontent.com/86117496/168828916-9b5d2b54-82e1-4f5d-ac1a-a37d47894dba.png)

After the changes:
![image](https://user-images.githubusercontent.com/86117496/168828781-b42a637f-5e79-496f-af3a-cbb17a53d590.png)

bobturneruk commented:
Good plan @SylviaWhittle - let us know if you have trouble running the tests on your machine. There are some great test / debug features in vscode.

SylviaWhittle commented:
I've had a try, but I have been unable to find how to run the tests. The readme says to `pip install ".[tests]"`, which I did and was successful, however there doesn't appear to be instruction on where to go from here. I have tried running various files in testing and also running topotracing but to no avail (it breaks saying 'no module named 'topostats.logs'') 

Sorry to be a bother @bobturneruk

ns-rse reviewed (commented):
Great work, as you've already noted the CSV against which the returned statistics is compared will need updating and if you're able to add a test for `calculate_points()` that would be great too.

SylviaWhittle commented:
@ns-rse @bobturneruk I have managed to run the tests, and I have replaced the old CSV file, however this did not work. I have manually checked that the grainstats output file from running topostats is the same as the file that I save in the resources folder, however the test is still saying things are different. Any ideas? 

The error
![image](https://user-images.githubusercontent.com/86117496/168860723-9b2c693a-9644-4dfe-9555-accdcd069cb6.png)

My comparison of the two csv files
![image](https://user-images.githubusercontent.com/86117496/168860834-29f95bbd-268b-41f3-b4fb-4019255ea781.png)

SylviaWhittle commented on topostats/topotracing.py:
```diff
@@ -212,7 +212,7 @@ def process_scan(
     # Calculate grain statistics
     grainstats = GrainStats(
-        data=zero_averaged_background,
+        data=gaussian_filtered,
```
I would say that we might want to keep it since it is a good visual check to see how well the flattening has worked. @alicepyne will have views.

bobturneruk commented:
![image](https://user-images.githubusercontent.com/34244196/168989844-651c582d-701e-4776-b1c3-b3da56587820.png)
FYI this is how I see the test in vscode.

bobturneruk commented:
I can see from this that the dataframe being made by the software has a different shape to that in the reference (2 extra columns, I think).

SylviaWhittle commented:
I think you are referring to the version that GitHub tested. That version results in a failed test expectedly because I had not yet updated the CSV file. On my local (which I haven't pushed yet), I have replaced the CSV file by copying the output grainstats.csv and pasting it into the testing resources folder, renaming it to `minicircle_grainstats_20220517.csv`, however when I run pytest locally, it still fails.
@bobturneruk

bobturneruk commented:
Please push your new version of the `csv` (we can always change back).

(I'm struggling to get the test debugging I was so chuffed with to work!).

SylviaWhittle commented:
I think I must have omitted to do something or somehow messed up checking the CSV files are the same (though I checked multiple times in google sheets, and also by re-running topotracing and copying the output csv file..)

bobturneruk commented:
Thanks for pushing. I have some meetings now but will try to get back to this today. I'm sure having the debugger working will help!

bobturneruk commented:
Ah yes, I now see the two DFs are the same shape but have different values.

SylviaWhittle commented:
They do not have different values. I am sure of this. 

The test fails because the testing script reads the `height_min` columns as being different:

![image](https://user-images.githubusercontent.com/86117496/168994256-00460291-fe14-41de-918b-87e5552a8987.png)

However, I manually ran the code just now and opened the CSV files and found the height_min values to be equal:

Output from running topotracing:
![image](https://user-images.githubusercontent.com/86117496/168994499-77bf1f97-44e0-44c9-835c-022e32984f66.png)

`minicircle_grainstats_20220517.csv`:
![image](https://user-images.githubusercontent.com/86117496/168994606-eb53345f-3a01-4f5a-8795-fb28d844a7d8.png)

In the failed test, the LHS values are not equal to this. I searched for the top value in my repo and found that these values are from the old `minicircle_grainstats_20220509.csv` file. But I do not know how it could access these values since I renamed `minicircle_grainstats_22020509.csv` to `minicircle_grainstats_20220505_OLD.csv` and I changed all references in the code to be the new file, `minicircle_grainstats_20220517.csv`. Perhaps it isn't pulling these old values from file, and instead somehow the code is producing them despite my changes?

@bobturneruk @ns-rse

bobturneruk commented:
So does this mean the test is generating different values to `topotracing`?

bobturneruk commented:
There is a difference in the data being used to create the GrainStats in `conftest.py` (https://github.com/AFM-SPM/TopoStats/blob/f8dcbc3d005d6fe04245043cd085579f7cacdbb3/tests/conftest.py#L293) and in `topotracing.py` (https://github.com/AFM-SPM/TopoStats/blob/f8dcbc3d005d6fe04245043cd085579f7cacdbb3/topostats/topotracing.py#L214). 

Perhaps different flattening? I think either `minicircle_zero_average_background` in `conftest.py` or `gaussian_filtered` in `topotracing.py` need to be looked at. There may be other factors, too.

SylviaWhittle commented:
@bobturneruk Well spotted! @ns-rse and I just had a call to talk about it and @ns-rse found it too 😄 Latest commit fixes it (hopefully). Thank you both!

bobturneruk commented:
Glad you both got it - I enjoyed the journey. Nice seeing those tests all pass!
